internal/controller/storageclient_controller.go

```diff
@@ -309,26 +416,125 @@ func (r *StorageClientReconciler) reconcilePhases() (ctrl.Result, error) {
 			}
-		}
-	}
-	if r.storageClient.GetAnnotations()[storageClaimProcessedAnnotationKey] != "true" {
-		if err := r.reconcileBlockStorageClaim(); err != nil {
-			return reconcile.Result{}, err
-		}
+		case "StorageClass":
+			storageClass := &storagev1.StorageClass{}
+			storageClass.Name = eResource.Name
+			if err := utils.CreateOrReplace(r.ctx, r.Client, storageClass, func() error {
+				if err := json.Unmarshal(eResource.Data, &storageClass); err != nil {
+					return fmt.Errorf("failed to unmarshal storageclass configuration response: %v", err)
+				}
+				// ensure name even if it was overwritten by unmarshal of whole resource
+				storageClass.Name = eResource.Name
+				removeStorageClaimAsOwner(storageClass)
+				if err := r.own(storageClass); err != nil {
+					return fmt.Errorf("failed to own Storage Class resource: %v", err)
+				}
 
-		if err := r.reconcileSharedfileStorageClaim(); err != nil {
-			return reconcile.Result{}, err
-		}
+				if storageClass.Parameters == nil {
+					storageClass.Parameters = map[string]string{}
+				}
+				storageClass.Parameters["csi.storage.k8s.io/provisioner-secret-namespace"] = r.OperatorNamespace
+				storageClass.Parameters["csi.storage.k8s.io/node-stage-secret-namespace"] = r.OperatorNamespace
+				storageClass.Parameters["csi.storage.k8s.io/controller-expand-secret-namespace"] = r.OperatorNamespace
 
-		utils.AddAnnotation(r.storageClient, storageClaimProcessedAnnotationKey, "true")
-		if err := r.update(r.storageClient); err != nil {
-			return reconcile.Result{}, fmt.Errorf("failed to update StorageClient with claim processed annotation: %v", err)
-		}
-	}
+				utils.AddLabels(storageClass, eResource.Labels)
+				utils.AddAnnotations(storageClass, eResource.Annotations)
+				return nil
+			}); err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to create or update StorageClass: %s", err)
+			}
+		case "VolumeSnapshotClass":
+			snapshotClass := &snapapi.VolumeSnapshotClass{}
+			snapshotClass.Name = eResource.Name
+			if err := utils.CreateOrReplace(r.ctx, r.Client, snapshotClass, func() error {
+				if err := json.Unmarshal(eResource.Data, snapshotClass); err != nil {
+					return fmt.Errorf("failed to unmarshal snapshotclass configuration response: %v", err)
+				}
+				snapshotClass.Name = eResource.Name
+				removeStorageClaimAsOwner(snapshotClass)
+				if err := r.own(snapshotClass); err != nil {
+					return fmt.Errorf("failed to own VolumeSnapshotClass resource: %v", err)
+				}
 
-	if utils.AddAnnotation(r.storageClient, utils.DesiredConfigHashAnnotationKey, storageClientResponse.DesiredConfigHash) {
-		if err := r.update(r.storageClient); err != nil {
-			return reconcile.Result{}, fmt.Errorf("failed to update StorageClient with desired config hash annotation: %v", err)
+				if snapshotClass.Parameters == nil {
+					snapshotClass.Parameters = map[string]string{}
+				}
+				snapshotClass.Parameters["csi.storage.k8s.io/snapshotter-secret-namespace"] = r.OperatorNamespace
```

Comment:
Why are we adding these parameters here and not on the provider?

Reply:
#323 is the client side fix and provider side will be done at a later point.